<a href="https://colab.research.google.com/github/arora123/Analytics-DS-ML-Projects-for-beginners_Python/blob/main/HR_Analytics__Project_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project Information

It is a very basic analytics project for beginners to understand how to explore data to get some insights using python

- Library used - Pandas

- Data - https://raw.githubusercontent.com/arora123/Data/master/emp-data.csv

- Data Description - Name, Gender, Department, Salary, Location & Ratings of 1015 employees [All column names are self explainatory]

- Objectives - 
1. How many Males/Females are there in the entire organization? 
2. How many Males/Females are there in the each department or for each location? 
3. For which department is the average Pay highest?
4. For which location is the average Pay highest?
5. What percentage of employees received good & very good rating? What about poor & very poor rating? and average rating?
6. Compute gender pay gap for each department. Interpret
7. Compute gender pay gap for each location. Interpret



# Loading Data

In [50]:
import pandas as pd

# emp = pd.read_csv("D:\Learning\Learn Excel\HR Analytics\emp-data.csv")
emp = pd.read_csv("https://raw.githubusercontent.com/arora123/Data/master/emp-data.csv")


emp.head()

Name  Gender   Department     Salary          Loc     Rating
0      Ches Bonnell    Male        Sales   $88,050      Bellevue  Very Good
1  Garwin Peasegood  Female  Engineering   $68,220      Bellevue       Good
2   Sidoney Yitzhok  Female          NaN  $118,440    Wellington  Not Rated
3  Saunders Blumson     NaN        Legal   $56,370   Los Angeles  Very Good
4    Gardy Grigorey  Female      Support  $107,090   Los Angeles       Poor

Did you notice, values in Salary column are containing '$' and ',' ?

These values might be stored as string data type.

Let's check:

# Exploring Data

In [51]:
emp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        1015 non-null   object
 1   Gender      972 non-null    object
 2   Department  987 non-null    object
 3   Salary      972 non-null    object
 4   Loc         1015 non-null   object
 5   Rating      1015 non-null   object
dtypes: object(6)
memory usage: 47.7+ KB


**Observations:**

1. All columns including 'Salary' are of 'object' data type, while statistically 'Salary' should be of numeric (float to be specific) data type
2. There are null values in Department, Gender & Salary columns

We will handle these issues one by one.

## let's check data type for all values in 'Salary' column

In [52]:
# emp['Salary'].apply(type)
emp['Salary'].apply(type).value_counts()

<class 'str'>      972
<class 'float'>     43
Name: Salary, dtype: int64

'Salary' column contains a mix of string & float values

To convert 'Salary' column to float data type, we need to first remove dollar sign & comma and then convert it to float (typecasting)

## Let's remove dollar sign & comma from 'Salary' column
There are multiple approaches to do the same. 

https://pbpython.com/currency-cleanup.html

In [53]:
# Using str.replace method to remove dollar sign & comma
emp['Salary'] = emp['Salary'].str.replace(',', '')
emp['Salary'] = emp['Salary'].str.replace('$', '')
print('After removing unwanted strings from Salary column: ', '\n', emp['Salary'].head(), '\n')

emp['Salary'] = emp['Salary'].astype(float)
print('Converting to float data type Salary column: ', '\n', emp['Salary'].head())

# we can also perform these operations in one line of code
# emp['Salary'] = emp['Salary'].str.replace(',', '').str.replace('$', '').astype(float)

After removing unwanted strings from Salary column:  
 0     88050 
1     68220 
2    118440 
3     56370 
4    107090 
Name: Salary, dtype: object 

Converting to float data type Salary column:  
 0     88050.0
1     68220.0
2    118440.0
3     56370.0
4    107090.0
Name: Salary, dtype: float64


## Other approaches

In [54]:
# Using a function which checks data type of each value in a column in remove unwanted strings 

# def clean_currency(x):
#     """ If the value is a string, then remove currency symbol and delimiters
#     otherwise, the value is numeric and can be converted
#     """
#     if isinstance(x, str):
#         return(x.replace('$', '').replace(',', ''))
#     return(x)

# emp['Salary'] = emp['Salary'].apply(clean_currency).astype(float)


In [55]:
# Using regex

# emp['Salary'] = emp['Salary'].replace({'\$': '', ',': ''}, regex=True).astype(float)


##Let's now understand missing values


In [56]:
 # All rows with missing values in any column
emp[emp.isna().any(axis=1)]

Name  Gender  ...          Loc     Rating
2      Sidoney Yitzhok  Female  ...   Wellington  Not Rated
3     Saunders Blumson     NaN  ...  Los Angeles  Very Good
8         Rayna Gamlin     NaN  ...   Wellington    Average
30       Mollie Hanway    Male  ...     Bellevue    Average
45         Fancy Bonin     NaN  ...     Bellevue       Good
...                ...     ...  ...          ...        ...
974   Jerrilee Maginot    Male  ...  Los Angeles    Average
980      Jolynn Edkins  Female  ...  Los Angeles    Average
989   Frasquito Mosley     NaN  ...   Wellington  Not Rated
992    Fairfax Wallsam     NaN  ...   Wellington    Average
1005  Madge McCloughen     NaN  ...  Los Angeles    Average

[109 rows x 6 columns]

In [57]:
# Rows with missing values in Salary & Department column
emp[(emp['Salary'].isna()) & (emp['Department'].isna())]

Name  Gender Department  Salary       Loc   Rating
343     Dewey Berthod  Female        NaN     NaN  Bellevue  Average
470  Aloise MacCathay    Male        NaN     NaN  Bellevue  Average

Only two rows with missing values in both column, may be they've left

 - Missing Salary might be indicating employees who have left, we can remove those rows
 - Missing values in department might be indicating, they have moved to sister company, we will remove these rows too

## Removing missing values from 'Salary' & 'Department' column 

In [58]:
 emp.dropna(subset = ['Salary', 'Department'], inplace=True)

In [59]:
emp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 946 entries, 0 to 1014
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Name        946 non-null    object 
 1   Gender      906 non-null    object 
 2   Department  946 non-null    object 
 3   Salary      946 non-null    float64
 4   Loc         946 non-null    object 
 5   Rating      946 non-null    object 
dtypes: float64(1), object(5)
memory usage: 51.7+ KB


Now, the missing values in 'Gender' column might be those who've chosen not to reveal their gender
Or they didn't find appropriate category

## Let's fill missing values in 'Gender' column by 'Not revealed' or 'Unknown' or 'Other'

In [60]:
emp['Gender'] = emp['Gender'].fillna('Not Revealed')

## Let's look at descriptive statistics for 'all' columns

In [61]:
emp.describe(include = 'all') 

Name Gender  ...          Loc   Rating
count            946    946  ...          946      946
unique           874      3  ...            3        6
top     Orton Livick   Male  ...  Los Angeles  Average
freq               2    465  ...          361      420
mean             NaN    NaN  ...          NaN      NaN
std              NaN    NaN  ...          NaN      NaN
min              NaN    NaN  ...          NaN      NaN
25%              NaN    NaN  ...          NaN      NaN
50%              NaN    NaN  ...          NaN      NaN
75%              NaN    NaN  ...          NaN      NaN
max              NaN    NaN  ...          NaN      NaN

[11 rows x 6 columns]

**Observations:**

1. Name column - out of 946 entries, only 874 are unique. It indicates presence of duplicate entries like the name 'Calvin O'Carroll' occuring twice
2. Gender Column - three categories, Male, Female & 'Not Revealed' with more (465) Males in the data set
3. 12 categories of departments, 3 locations and 6 categories of ratings (maximum employees got ;Average' rating)
4. Salary - mean value is 73703 and median is 73300. It indicates right skewed distribution as expected


## Let's explore duplicate entries in 'Name' column

In [62]:
#let's look at those names
# emp.duplicated('Name')
# emp.duplicated('Name').value_counts() # 72 names are duplicated
emp[emp.duplicated('Name') == True] # 72 names are duplicated

Name  Gender  ...          Loc     Rating
130         Larry Pioch    Male  ...   Wellington       Good
145   Adella Hartshorne  Female  ...   Wellington       Good
155         Katya Hundy    Male  ...  Los Angeles       Poor
168       Crawford Scad    Male  ...     Bellevue    Average
272    Calvin O'Carroll  Female  ...  Los Angeles  Very Poor
...                 ...     ...  ...          ...        ...
981      Beverie Moffet  Female  ...  Los Angeles    Average
983        Alida Welman    Male  ...     Bellevue       Poor
996      Joella Maevela  Female  ...  Los Angeles       Good
1004     Curtice Advani    Male  ...     Bellevue       Good
1008         Anni Dinse    Male  ...     Bellevue  Not Rated

[72 rows x 6 columns]

In [63]:
#Another approach
names = emp.Name.value_counts()
names[names > 1]

Orton Livick                2
Rhiamon Mollison            2
Murry Dryburgh              2
Barbara-anne Kenchington    2
Joyce Esel                  2
                           ..
Sheff Gerdts                2
Joella Maevela              2
Win Arthurs                 2
Kissiah Maydway             2
Gilles Jaquet               2
Name: Name, Length: 72, dtype: int64

In [64]:
emp[emp.Name == 'Larry Pioch'] #same employee but rating different?
# emp[emp.Name == 'Adella Hartshorne'] #may be different employee or data entry error
# emp[emp.Name == "Calvin O'Carroll"] #may be different employee or data entry error

Name Gender                Department   Salary         Loc Rating
69   Larry Pioch   Male  Research and Development  49670.0  Wellington   Poor
130  Larry Pioch   Male  Research and Development  49670.0  Wellington   Good

In [65]:
#@title
# We can use excel to quickly watch rows with duplicated names for better understanding
# Let's write a csv file for these rows 
# emp1 = emp[emp.duplicated('Name') == True]

# emp1.to_csv('emp1.csv')
# in excel it shows no duplicates in emp and emp1 file?

**Observations:**
1. 72 names are repeated twice in the data set
2. Some entries look same and some look different

May be a few of those represent different employess with the same name but mostly it looks data entry error as salaries look same 

**How to Handle This Issue?:** 

Ideally, as an analyst, you need to verify the record of these employess from employee database 

Usually in an organizations, each employee is given a unique employee-id which helps to identify the issues related to duplicate names

Such data quality issues can have big impact on analysis

Here, for the purpose of demostration, we are removing all rows with duplicated names 

## Let's remove all rows with duplicated names 

In [66]:
emp.drop_duplicates(subset = 'Name', inplace=True) # default keep='first'
# Here, we are dropping all duplicates except for the first occurrence. 
# Other options are keep='last' or keep='False'
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html

In [67]:
emp.describe(include='all')

Name Gender  ...          Loc   Rating
count               874    874  ...          874      874
unique              874      3  ...            3        6
top     Blythe Clipston   Male  ...  Los Angeles  Average
freq                  1    430  ...          337      400
mean                NaN    NaN  ...          NaN      NaN
std                 NaN    NaN  ...          NaN      NaN
min                 NaN    NaN  ...          NaN      NaN
25%                 NaN    NaN  ...          NaN      NaN
50%                 NaN    NaN  ...          NaN      NaN
75%                 NaN    NaN  ...          NaN      NaN
max                 NaN    NaN  ...          NaN      NaN

[11 rows x 6 columns]

- Now, we've 874 rows
- 'Salary' is still right skewed as expected

# Getting Insights

## How many Males/Females are there in the entire organization? 


In [68]:
emp['Gender'].value_counts()
# emp['Gender'].value_counts().sum() #874

Male            430
Female          406
Not Revealed     38
Name: Gender, dtype: int64

In [69]:
Gender_perc = emp['Gender'].value_counts()/emp['Gender'].value_counts().sum()
Gender_perc.round(2)

Male            0.49
Female          0.46
Not Revealed    0.04
Name: Gender, dtype: float64

In [70]:
emp['Gender'].describe()

count      874
unique       3
top       Male
freq       430
Name: Gender, dtype: object

## Out of total 874 employees, 430 (49%) are males, 406 (46%) are females and Gender of 38 employees is not unknown to us  

## How many Males/Females are there in the each department or for each location? 

In [71]:
# Male/Female in each department
pd.crosstab(emp['Gender'], emp['Department'])

Department    Accounting  Business Development  ...  Support  Training
Gender                                          ...                   
Female                27                    38  ...       30        34
Male                  36                    33  ...       41        35
Not Revealed           2                     3  ...        4         3

[3 rows x 12 columns]

## We have more females in 'Business Development', 'HR' and 'Research & Development'
In rest all departments, we've more males than females

Here, we are ignoring gender of those who've chosen not to reveal


In [72]:
# Male/Female in each department
pd.crosstab(emp['Gender'], emp['Loc'])

Loc           Bellevue  Los Angeles  Wellington
Gender                                         
Female             111          151         144
Male               114          172         144
Not Revealed         8           14          16

## We've more males than females at all locations except Wellington, where we've equal number of males & females.

## For which department is the average Pay highest? And how much is it?

In [73]:
# Department-wise average salary
emp.groupby('Department')['Salary'].mean().round(0)

Department
Accounting                  75896.0
Business Development        76309.0
Engineering                 71959.0
Human Resources             71473.0
Legal                       72663.0
Marketing                   76646.0
Product Management          72082.0
Research and Development    69768.0
Sales                       71751.0
Services                    76711.0
Support                     74823.0
Training                    76266.0
Name: Salary, dtype: float64

In [74]:
emp.groupby('Department')['Salary'].mean().round(0).max()
emp.groupby('Department')['Salary'].mean().round(0).min()

69768.0

## Highest average salary is $76711 for 'Services'  department and Lowest is $ 69768 for 'Research and Development' department

## Which location has highest & lowest average salary?

In [75]:
# Location-wise average salary
emp.groupby('Loc')['Salary'].mean().round(0)

Loc
Bellevue       76100.0
Los Angeles    73580.0
Wellington     72372.0
Name: Salary, dtype: float64

## Wellington location provides lowest average salary ($72372) & Bellevue provides highest average salary ($76100)

## What percentage of employees received good & very good rating? What about poor & very poor rating? and average rating?

In [76]:
emp['Rating'].value_counts()

Average      400
Good         163
Poor         116
Very Good     81
Not Rated     61
Very Poor     53
Name: Rating, dtype: int64

In [77]:
rating_perc = emp['Rating'].value_counts()/emp['Rating'].value_counts().sum()
100*rating_perc.round(2)

Average      46.0
Good         19.0
Poor         13.0
Very Good     9.0
Not Rated     7.0
Very Poor     6.0
Name: Rating, dtype: float64

## Compute gender pay gap across the organization. Interpret

In [78]:
# Gender-wise average salary
emp.groupby('Gender')['Salary'].mean().round(0)

Gender
Female          72610.0
Male            74507.0
Not Revealed    79245.0
Name: Salary, dtype: float64

In [79]:
pay_m = emp[emp['Gender'] == 'Male']['Salary'].mean()
print("Average pay for males across the organization: ", round(pay_m, 0), '\n')

pay_f = emp[emp['Gender'] == 'Female']['Salary'].mean()
print("Average pay for females across the organization: ", round(pay_f, 0))

Average pay for males across the organization:  74507.0 

Average pay for females across the organization:  72610.0


Data suggests that on an average females get less salary than males.

## Let's compute gender pay gap across organization

In [80]:
pay_gap = (pay_m/pay_f - 1)*100 
pay_gap 

2.6124554508039743

**Interpretation:**

- positive value indicate females are underpaid

if this value is negative, it would indicate that males are underpaid

## Compute gender pay gap for each department. Interpret

In [81]:
# Just exploring department-wise average salary
emp.groupby('Department')['Salary'].mean().round(0)

Department
Accounting                  75896.0
Business Development        76309.0
Engineering                 71959.0
Human Resources             71473.0
Legal                       72663.0
Marketing                   76646.0
Product Management          72082.0
Research and Development    69768.0
Sales                       71751.0
Services                    76711.0
Support                     74823.0
Training                    76266.0
Name: Salary, dtype: float64

In [82]:
# Minimum average salary for which department?
emp.groupby('Department')['Salary'].mean().round(0).min()
# for research & development

69768.0

In [83]:
# Maximum average salary for which department?
emp.groupby('Department')['Salary'].mean().round(0).max()  
# for business development

76711.0

In [84]:
# Department-wise average salary for females & males
emp.groupby(['Department', 'Gender'])['Salary'].mean().round(0)  

# emp.groupby(['Gender', 'Department'])['Salary'].mean().round(0)  

Department                Gender      
Accounting                Female           72814.0
                          Male             76726.0
                          Not Revealed    102560.0
Business Development      Female           74667.0
                          Male             80596.0
                          Not Revealed     49957.0
Engineering               Female           74868.0
                          Male             68097.0
                          Not Revealed     78626.0
Human Resources           Female           67062.0
                          Male             74345.0
                          Not Revealed     94283.0
Legal                     Female           69904.0
                          Male             73832.0
                          Not Revealed     81875.0
Marketing                 Female           79108.0
                          Male             73242.0
                          Not Revealed    105870.0
Product Management        Female           

We can not make quick conclusions looking at the table directly 

## Let's compute average salary for males & females in each department

In [85]:
#Creating separate data sets of Males & Females
males = emp[emp['Gender'] == 'Male']
print(males.shape) # 430 rows as expected

females = emp[emp['Gender'] == 'Female']
females.shape # 430 rows as expected

(430, 6)


(406, 6)

In [86]:
# Male's average pay for each department

pay_m_dept = males.groupby('Department')['Salary'].mean().round(0)
pay_m_dept 

Department
Accounting                  76726.0
Business Development        80596.0
Engineering                 68097.0
Human Resources             74345.0
Legal                       73832.0
Marketing                   73242.0
Product Management          73484.0
Research and Development    70338.0
Sales                       71287.0
Services                    80379.0
Support                     77054.0
Training                    74247.0
Name: Salary, dtype: float64

In [87]:
# Female's average pay for each department

pay_f_dept = females.groupby('Department')['Salary'].mean().round(0)
pay_f_dept 

Department
Accounting                  72814.0
Business Development        74667.0
Engineering                 74868.0
Human Resources             67062.0
Legal                       69904.0
Marketing                   79108.0
Product Management          71558.0
Research and Development    67888.0
Sales                       70144.0
Services                    74161.0
Support                     71586.0
Training                    77579.0
Name: Salary, dtype: float64

In [89]:
# Department-wise gender pay gap
pay_gap_dept = ((pay_m_dept/pay_f_dept) - 1)*100
round(pay_gap_dept, 0)

Department
Accounting                   5.0
Business Development         8.0
Engineering                 -9.0
Human Resources             11.0
Legal                        6.0
Marketing                   -7.0
Product Management           3.0
Research and Development     4.0
Sales                        2.0
Services                     8.0
Support                      8.0
Training                    -4.0
Name: Salary, dtype: float64

## Using a function to identify positive or negative wage gap

read more about data frame styling: https://pandas.pydata.org/pandas-docs/stable/user_guide/style.html

In [95]:
def positive_negative(val):
    """
    Takes a scalar and returns a string with
    the css property `'color: red'` for negative
    strings, black otherwise.
    """
    color = 'red' if val < 0 else 'green'
    return 'color: %s' % color

In [96]:
gap_dept = pd.DataFrame(pay_gap_dept)
gap_dept.style.applymap(positive_negative)

**Observations:**
- In departments like Engineering, Marketing & Training, female's average pay is higher than that of males. 
- For rest of the departments, Male's get more than females.

We should also explore distribution of salaries, experience or other factors for better understanding

## Using a function to highlight maximum value

In [99]:
def highlight_max(s):
    '''
    highlight the maximum in a Series yellow.
    '''
    is_max = s == s.max()
    return ['background-color: blue' if v else '' for v in is_max]

In [100]:
gap_dept.style.apply(highlight_max)

Observations:
- Highest wage gap is in HR department where Male's average pay is almost 11% (10.86) higher than that of females

( 74345.0 - 67062)/74345.0 = 0.09796220324
- Alternately, we can say that Female's average pay is almost 10% (9.796) lower than that of Males


In [40]:
#Location-wise average salary for females & males

emp.groupby(['Loc', 'Gender'])['Salary'].mean().round(0)  

Loc          Gender      
Bellevue     Female          75103.0
             Male            76315.0
             Not Revealed    86860.0
Los Angeles  Female          72568.0
             Male            74056.0
             Not Revealed    78643.0
Wellington   Female          70732.0
             Male            73613.0
             Not Revealed    75965.0
Name: Salary, dtype: float64

In [44]:
# Male's average pay for each Location

pay_m_loc = males.groupby('Loc')['Salary'].mean().round(0)
pay_m_loc 


Loc
Bellevue       76315.0
Los Angeles    74056.0
Wellington     73613.0
Name: Salary, dtype: float64

In [45]:
# Female's average pay for each Location

pay_f_loc = females.groupby('Loc')['Salary'].mean().round(0)
pay_f_loc 

Loc
Bellevue       75103.0
Los Angeles    72568.0
Wellington     70732.0
Name: Salary, dtype: float64

In [46]:
# Location-wise gender pay gap

pay_gap_loc = ((pay_m_loc/pay_f_loc) - 1)*100
pay_gap_loc

Loc
Bellevue       1.613784
Los Angeles    2.050491
Wellington     4.073121
Name: Salary, dtype: float64

## At all locations, males make more money than their female counterparts.
## Highest wage gap is at Wellington location where male's average salary is 4.07 % higher than female's average salary